### PR TITLE
[IMP] website: added a conditional visibility option on <section> blocks

### PR DESCRIPTION
**Pull request by detrouxdev**
Provide the user with an option to hide or display a section tag
depending on the visitor's location (if geoip enabled), language,
utm_medium, utm_source and utm_campaign.

User can use a M2M widget to select different records. These will be
used to create a CSS Selector on save that will hide the section tag
depending on the various options they selected.
This CSS selectors will be applied, as well as various data attributes
when the page loads to hide the targeted tag. All of these operations
are handled client side.

The geoip country had to be added to the session for the feature to work
with countries.

task-2381049

### CODE REVIEW

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 1]
```diff
@@ -2080,8 +2084,8 @@ const ListUserValueWidget = UserValueWidget.extend({
                 return;
             }
             prepare();
-            const { id, display_name } = JSON.parse(widget.getMethodsParams('addRecord').recordData);
-            this._addItemToTable(id, display_name);
+            const { id, display_name, ...extra } = JSON.parse(widget.getMethodsParams('addRecord').recordData);
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:2087]: I don't really like to use this while it is still marked as an experimental feature: https://caniuse.com/mdn-javascript_operators_destructuring_rest_in_objects

File: addons/website/static/src/js/editor/snippets.options.js [hunk 2]
```diff
@@ -2739,6 +2739,194 @@ options.registry.ReplaceImage = options.Class.extend({
     },
 });
 
+options.registry.ConditionalVisibility = options.Class.extend({
+
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2743]: Remove this blank line and the blank line between functions. The docstring as separator is enough.

File: addons/website/static/src/js/editor/snippets.options.js [hunk 3]
```diff
@@ -2739,6 +2739,194 @@ options.registry.ReplaceImage = options.Class.extend({
...
+    /**
+     * We clear out the current style tag to make everything visible
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2745]: Try to add a docstring when you can on this PR and it should start with a congugated verb to complete the sentence "This function ...". So here "Clears out the...".

File: addons/website/static/src/js/editor/snippets.options.js [hunk 4]
```diff
@@ -2739,6 +2739,194 @@ options.registry.ReplaceImage = options.Class.extend({
...
+     * @override
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2746]: `@constructor` for the constructor and there should be a blank line before the first tag like this.

File: addons/website/static/src/js/editor/snippets.options.js [hunk 5]
```diff
@@ -2739,6 +2739,194 @@ options.registry.ReplaceImage = options.Class.extend({
...
+     */
+    init: function () {
+        const style = document.getElementById('wrap') && document.getElementById('wrap').getElementsByTagName('style')[0];
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2749]: If you can try using the convention to put "El" at the end of variable containing html elements, I think it is a good idea. So here `styleEl`

File: addons/website/static/src/js/editor/snippets.options.js [hunk 6]
```diff
@@ -2739,6 +2739,194 @@ options.registry.ReplaceImage = options.Class.extend({
     },
 });
 
+options.registry.ConditionalVisibility = options.Class.extend({
+
+    /**
+     * We clear out the current style tag to make everything visible
+     * @override
+     */
+    init: function () {
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2748]: You can use ES6 syntax for all your method definitions:
```javascript
init() {
    // ...
},
```
(no function keyword)

File: addons/website/static/src/js/editor/snippets.options.js [hunk 7]
```diff
@@ -2739,6 +2739,194 @@ options.registry.ReplaceImage = options.Class.extend({
...
+            if (rottenIds) {
+                this.selectValue(false, widget._value, widget._methodsParams);
+            }
+        });
+    },
+
+    /**
+     * @see this.selectClass
+     */
+    selectValue(preivewMode, widgetValue, params) {
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2786]: Typo, try to re-review your whole PR to make sure there is not any other :wink:

File: addons/website/static/src/js/editor/snippets.options.js [hunk 8]
```diff
@@ -2739,6 +2739,194 @@ options.registry.ReplaceImage = options.Class.extend({
...
+        const recordData = JSON.parse(widgetValue);
+        const ids = [];
+        const codes = [];
+        for (const record of recordData) {
+            ids.push(record.id);
+            codes.push(record[params.callWith]);
+        }
+        if (ids.length && codes.length) {
+            this.$target[0].dataset[params.attributeName + 'Id'] = ids.join(',');
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2795]: In case you don't know, we try to use literal strings to combine strings now. So
```javascript
`${params.attributeName}Id`
```

File: addons/website/static/src/js/editor/snippets.options.js [hunk 9]
```diff
@@ -2739,6 +2739,194 @@ options.registry.ReplaceImage = options.Class.extend({
...
+
+    /**
+     * @override
+     */
+    selectDataAttribute(preivewMode, widgetValue, params) {
+        this._super(...arguments);
+        if (params.attributeName === 'visibility' && widgetValue === 'conditional') {
+            const collapse = this.$el.children('we-collapse')[0];
+            collapse.classList.add('active');
+            $(collapse).children('.o_we_collapse_toggler')[0].classList.add('active');
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2811]: There is code in charge to handle the opening of collapse elements, you should not duplicate its logic.

File: addons/website/static/src/js/editor/snippets.options.js [hunk 10]
```diff
@@ -2739,6 +2739,194 @@ options.registry.ReplaceImage = options.Class.extend({
...
+     */
+    onTargetShow: async function () {
+        this.$target[0].style = this.$target[0].previousStyle;
+    },
+
+    /**
+     * @override
+     */
+    onTargetHide: async function () {
+        this.$target[0].previousStyle = this.$target.style;
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2826]: A style attribute on the jQuery object? Seems strange

File: addons/website/static/src/js/editor/snippets.options.js [hunk 11]
```diff
@@ -2739,6 +2739,194 @@ options.registry.ReplaceImage = options.Class.extend({
...
+        delete this.$target[0].dataset['invisible'];
+        if (this.$target[0].dataset['visibility'] !== 'conditional') {
+            delete this.$target[0].dataset['visibilityId'];
+            return this._super(...arguments);
+        }
+        let visibilityId = '';
+        const onlyAttributes = [];
+        const hideAttributes = [];
+        for (const attr of this.$target[0].attributes) {
+            if (attr.name.startsWith('data-visibility-') && !attr.name.startsWith('data-visibility-rule') && !attr.name.endsWith('-id')) {
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2845]: Should not you loop through the dataset of the element instead of its attributes?

File: addons/website/static/src/js/editor/snippets.options.js [hunk 12]
```diff
@@ -2739,6 +2739,194 @@ options.registry.ReplaceImage = options.Class.extend({
...
+    // Private
+    //--------------------------------------------------------------------------
+
+    /**
+     * Add a "display: none;" rule with the corresponding selector to a style tag in the #wrap of the page
+     * @param {string} selector a CSS selector
+     */
+
+    _applyStyleToDom: function (selector) {
+        const wrap = document.getElementById('wrap');
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2896]: It is not guaranteed there is a wrap in each page (even if it is probably the case) but overal, the wrap element is not necessarily the element you are editing. If the user edits a product description, the content will be saved in that model field... so should the associated style tag. You can probably find the editable element with `this.$target.closest('.o_editable')` ... but maybe to review in a few days when the new editor is merged.

File: addons/website/tests/test_ui.py [hunk 13]
```diff
@@ -192,3 +192,16 @@ def test_07_snippet_version(self):
...
+    def test_09_website_conditional_visibility(self):
+        self.start_tour('/', 'conditional_visibility', login='admin')
+        self.browser_js('/?utm_medium=Email', """
+                const el = document.getElementById('wrap').getElementsByClassName('s_text_image')[0]
+                const style = window.getComputedStyle(el)
+                if (style.display === 'none') {
+                    console.error('this element should be visible now')
+                } else {
+                    console.log('test successful')
+                }
```
Comment (qsm-odoo) [inline addons/website/tests/test_ui.py:205]: Even if it feels overkill, I would prefer no JS in a python file: create a tour with an unique step just for this check.

File: addons/website/views/assets.xml [hunk 14]
```diff
@@ -62,6 +63,12 @@
         <link rel="stylesheet" type="text/scss" href="/website/static/src/scss/bootstrap_overridden.scss"/>
     </xpath>
 </template>
+ 
+<template id="_assets_frontend_minimal_js" inherit_id="web._assets_frontend_minimal_js">
+    <xpath expr="//script[last()]" position="after">
+        <script type="text/javascript" src="/website/static/src/js/content/inject_dom.js"/>
+    </xpath>
+</template>
```
Comment (qsm-odoo) [inline addons/website/views/assets.xml:71]: You can rebase your work to use the new system of assets :grimacing:

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
I did not check in details yet but here is a first batch of comments. Can't be sure yet, but this seems like a good work :+1: Try to apply a comment I add somewhere on the whole PR (not only where I comment :wink:)

Try to reorganize your github history too. When marking the task as finished and ready for review, the commits should normally be already clean: here I don't know if the work can be splitted in multiple commits (I don't think so?) so maybe there should only be one commit with a proper commit message :+1: 

Courage! :smile:

File: addons/website/static/src/js/editor/snippets.options.js (hunk 11 above)
Comment (detrouxdev) [reply addons/website/static/src/js/editor/snippets.options.js:2845]: Because we later on add those attributes as "data-attr-name" in a CSS rule, I thought it was easier to loop through the attributes to get their names in the correct format. I could loop on the dataset and replace the name, it would make the loop faster but would add the extra action of having to replace the string. I don't know which is faster.

**Comment by detrouxdev** [comment]
@qsm-odoo I think I resolved most of the initial issues and this task is ready for another review.

File: addons/website/static/tests/tours/conditional_visibility.js [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+        id: 's_text_image',
+        name: 'Text - Image',
+    },
+];
+tour.register('conditional_visibility_1', {
+    test: true,
+    url: '/',
+},
+[{
+    content: "enter eddit mode",
```
Comment (JKE-be) [inline addons/website/static/tests/tours/conditional_visibility.js:16]: dd

File: addons/website/models/website_page.py [hunk 16]
```diff
@@ -212,7 +212,7 @@ def _get_cache_key(self, req):
         # Always call me with super() AT THE END to have cache_key_expr appended as last element
         # It is the only way for end user to not use cache via expr.
         # E.g  (None if 'token' in request.params else 1,)  will bypass cache_time
-        cache_key = (req.website.id, req.lang, req.httprequest.path)
+        cache_key = (req.website.id, req.lang, req.session.geoip.get('country_code'), req.httprequest.path)
```
Comment (JKE-be) [inline addons/website/models/website_page.py:215]: what's your opinion about move this in session info @qsm-odoo  ? and avoid to fork cache ? 
This one has is removed in chm PR so it will be probably quickly broekn with their new cache...

File: addons/website/static/src/js/editor/snippets.options.js [hunk 17]
```diff
@@ -2822,6 +2822,165 @@ options.registry.ReplaceImage = options.Class.extend({
...
+    /**
+     * @override
+     */
+    async onTargetShow() {
+        this.$target[0].style = this.$target[0].previousStyle;
+    },
+
+
+    /**
+     * Reades the attributes of the traget and creates the selectors
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2909]: Reades :grimacing: Traget :grimacing: Not reviewed by yourself :wink:

File: addons/website/static/src/js/editor/snippets.options.js [hunk 18]
```diff
@@ -2822,6 +2822,165 @@ options.registry.ReplaceImage = options.Class.extend({
...
+     * Store them in data-attributes to then be reapplied by content/inject_dom.js
+     * @override
+    */
+    async cleanForSave() {
+        this.$target[0].style = this.$target[0].previousStyle;
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2914]: Not sure why this would be needed?

File: addons/website/static/src/js/editor/snippets.options.js [hunk 19]
```diff
@@ -2822,6 +2822,165 @@ options.registry.ReplaceImage = options.Class.extend({
...
+        delete this.$target[0].dataset.invisible;
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2915]: `dataset.invisible` is for elements which should always be invisible if I remember well, so I don't think there should even be one here?

File: addons/website/static/src/js/editor/snippets.options.js [hunk 20]
```diff
@@ -2822,6 +2822,165 @@ options.registry.ReplaceImage = options.Class.extend({
...
+        if (this.$target[0].dataset.visibility !== 'conditional') {
+            delete this.$target[0].dataset.visibilityId;
+            delete this.$target[0].dataset.visibilityRule;
+            return this._super(...arguments);
+        }
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2920]: Does not seem to be the role of `cleanForSave`, the option which unset conditional visibility should delete the related attributes directly.

File: addons/website/views/website_templates.xml [hunk 21]
```diff
@@ -47,6 +47,7 @@
             'data-main-object': repr(main_object),
             'data-seo-object': repr(seo_object) if seo_object else None,
             'data-oe-company-name': res_company.name,
+            'data-country': request.session.geoip.get('country-code'),
```
Comment (qsm-odoo) [inline addons/website/views/website_templates.xml:50]: So as said above: not on the HTML directly anymore but in the session_info

File: addons/website/static/src/js/editor/snippets.options.js [hunk 22]
```diff
@@ -2822,6 +2822,165 @@ options.registry.ReplaceImage = options.Class.extend({
...
+                const shortName = attr.name.replace('data-visibility-value-', '');
+                let dataName, values;
+                if (shortName === 'lang') {
+                    dataName = 'lang';
+                    values = attr.value.replace(/_/g, '-').split(',');
+                } else {
+                    dataName = `data-${shortName}`;
+                    values = attr.value.split(',');
+                }
+                const hideFor = this.$target[0].attributes[`data-visibility-rule-${shortName}`] && this.$target[0].attributes[`data-visibility-rule-${shortName}`].value === 'hide';
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2935]: Don't repeat ```this.$target[0].attributes[`data-visibility-rule-${shortName}`]``` twice

File: addons/website/static/src/js/editor/snippets.options.js [hunk 23]
```diff
@@ -2822,6 +2822,165 @@ options.registry.ReplaceImage = options.Class.extend({
...
+                if (hideFor) {
+                    hideAttributes.push({ name: dataName, values: values });
+                } else {
+                    onlyAttributes.push({ name: dataName, values: values });
+                }
+                visibilityId += shortName + this.$target[0].attributes[`data-visibility-value-${shortName}-id`].value + (hideFor ? 'h' : '');
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2941]: I am not sure to understand this whole code, can you explain the different attributes you are reading? There seems to be data-visibility-value-XXX and data-visibility-value-XXX-id... and I don't see where and why is the difference.

File: addons/website/static/src/js/editor/snippets.options.js [hunk 24]
```diff
@@ -2822,6 +2822,165 @@ options.registry.ReplaceImage = options.Class.extend({
...
+            }
+        }
+        /*
+        * Creating CSS selectors based on those attributes, the reducers combine the attributes
+        * Important to note how the CSS Rule works : We hide the content that match the selector. Even for "Visible to" option
+        * Therefore we hide when the HTML tag does not contain the values we're looking for
+        */
+        let selectors = '';
+        for (const attribute of onlyAttributes) {
+            // selector = htm:not([data-attr-1="valueAttr1"]):not([data-attr-1="valueAttr2"]) [data-visbility-id="ruleId"]
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2951]: typo

File: addons/website/static/src/js/editor/snippets.options.js [hunk 25]
```diff
@@ -2822,6 +2822,165 @@ options.registry.ReplaceImage = options.Class.extend({
...
+            selectors += selector + ', ';
+        }
+        for (const attribute of hideAttributes) {
+            // selector = html[data-attr-1="valueAttr1"] [data-visbility-id="ruleId"], html[data-attr-1="valueAttr2"] [data-visbility-id="ruleId"]
+            const selector = attribute.values.reduce((acc, cv, i, a) => acc += `html[${attribute.name}="${cv}"] [data-visibility-id="${visibilityId}"]${i !== a.length - 1 ? ',' : ''}`, '');
+            selectors += selector + ', ';
+        }
+        selectors = selectors.slice(0, -2);
+        if (selectors !== '') {
+            this.$target[0].dataset.visibilitySelectors = selectors;
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2962]: `if (selectors) {`

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Some comments on things to review and things you should explain so I can better review :+1: 

Also review your commit message:
- max 72 characters per line
- the title should complete the sentence "When applied, this commit will ..."
- No uppercase letter to start the tile
- No space before `:` in the title
- Add a blank line followed by exactly task-XXX where XXX is the ID of your task in the Odoo backend (you can see the ID in the URL on your task)

File: addons/website/static/src/js/content/inject_dom.js [hunk 26]
```diff
@@ -0,0 +1,21 @@
...
+        for (const name of cookieNames) {
+            const cookie = cookies[`odoo_${name}`];
+            if (cookie) {
+                htmlEl.setAttribute(`data-${name.replace('_', '-')}`, cookie);
+            }
+        }
+
+        const styleEl = document.createElement('style');
+        styleEl.id = "conditional_visibility";
+        document.getElementsByTagName('head')[0].appendChild(styleEl);
```
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/website/static/src/js/content/inject_dom.js:15]: It seems like to access styleEl.sheet, it needs to be added to the DOM first, maybe using it is unreliable then but I can't seem to find exactly when the sheet element is instantiated. Should I replace it with `styleEl.innerHTML += rule` ?

[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/website/static/src/js/content/inject_dom.js]: Mmh, strange, you can keep as you did :+1:

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 27]
```diff
@@ -1889,8 +1889,10 @@ const ListUserValueWidget = UserValueWidget.extend({
      * @private
      * @param {string || integer} id
      * @param {string} text
+     * @param {Object} recordData key, values that will be added
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:1892]: Optional params should be documented with the variable between `[]`. It's a mistake that `text` does not use that.

File: addons/website/static/src/js/editor/snippets.options.js [hunk 28]
```diff
@@ -2822,6 +2822,220 @@ options.registry.ReplaceImage = options.Class.extend({
...
+        {
+            selectName: 'visibilityValueUtmSource',
+            dataName: 'data-utm-source',
+        },
+        {
+            selectName: 'visibilityValueUtmCampaign',
+            dataName: 'data-utm-campaign',
+        },
+    ],
+    /**
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2848]: Blank line between properties and the first method

File: addons/website/static/src/js/editor/snippets.options.js [hunk 29]
```diff
@@ -2822,6 +2822,220 @@ options.registry.ReplaceImage = options.Class.extend({
...
+            this.trigger_up('snippet_option_visibility_update', { show: true });
+        }
+    },
+    /**
+     * Inserts or deletes record's id and value in target's data-attributes
+     * if no ids are selected, deletes the attribute
+     *
+     * @see this.selectClass for parameters
+     */
+    selectRecord(previewMode, widgetValue, params) {
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2865]: Check the organisation of options methods. There should be a "options" section comment before options and "lifecycle" method should be place before (onFocus, cleanForSave, etc).

File: addons/website/static/src/js/editor/snippets.options.js [hunk 30]
```diff
@@ -2822,6 +2822,220 @@ options.registry.ReplaceImage = options.Class.extend({
...
+        },
+    ],
+    /**
+     * @constructor
+     */
+    init() {
+        this._super(...arguments);
+        if (this.$target[0].dataset.visibility === 'conditional') {
+            // hack to use the same visibility pannel as the popups
+            this.$target[0].classList.add('o_snippet_invisible');
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2855]: I think the o_snippet_invisible should be added the same time as when the conditionnal visibility is added, I don't see why this should be (and would work?) if in the "init" method.

File: addons/website/static/src/js/editor/snippets.options.js [hunk 31]
```diff
@@ -2822,6 +2822,220 @@ options.registry.ReplaceImage = options.Class.extend({
...
+            dataName: 'data-utm-campaign',
+        },
+    ],
+    /**
+     * @constructor
+     */
+    init() {
+        this._super(...arguments);
+        if (this.$target[0].dataset.visibility === 'conditional') {
+            // hack to use the same visibility pannel as the popups
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2854]: Typo

File: addons/website/static/src/js/editor/snippets.options.js [hunk 32]
```diff
@@ -2822,6 +2822,220 @@ options.registry.ReplaceImage = options.Class.extend({
...
+     * content/inject_dom.js
+     *
+     * @override
+    */
+    async cleanForSave() {
+        // removes classes that were used for the invisible elements pannel
+        this.$target[0].classList.remove('o_snippet_invisible');
+        this.$target[0].classList.remove('d-none');
+        if (this.$target[0].dataset.visibility !== 'conditional') {
+            return await this._super(...arguments);
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2914]: `return await` or `return` doing the same thing in combination with `async` I would not write the `await` part. Seems also more natural to just "return super".

File: addons/website/static/src/js/editor/snippets.options.js [hunk 33]
```diff
@@ -2822,6 +2822,220 @@ options.registry.ReplaceImage = options.Class.extend({
...
+    /**
+     * Reads target's attributes and creates CSS selectors
+     * Stores them in data-attributes to then be reapplied by
+     * content/inject_dom.js
+     *
+     * @override
+    */
+    async cleanForSave() {
+        // removes classes that were used for the invisible elements pannel
+        this.$target[0].classList.remove('o_snippet_invisible');
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2911]: Ok, not really needed IMO. The snippet is conditionally visible, it can be marked as such forever.

File: addons/website/static/src/js/editor/snippets.options.js [hunk 34]
```diff
@@ -2822,6 +2822,220 @@ options.registry.ReplaceImage = options.Class.extend({
...
+        this.$target[0].classList.remove('d-none');
+        if (this.$target[0].dataset.visibility !== 'conditional') {
+            return await this._super(...arguments);
+        }
+        /**
+         * There are 2 data attributes per option
+         * One that stores the current records selected
+         * Another that stores the value of the rule "Hide for / Visible for"
+         */
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2920]: We normally only use `/** */` style comments for method/class documentation, you can use a series of one line comments for those.

File: addons/website/static/src/js/editor/snippets.options.js [hunk 35]
```diff
@@ -2822,6 +2822,220 @@ options.registry.ReplaceImage = options.Class.extend({
...
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+
+    /**
+     * @override
+     */
+    async _computeWidgetState(methodName, params) {
+        if (methodName === 'selectRecord') {
+            // m2m value widget does not use computeWidgetState
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2997]: Why is that actually? Something related to @sdegueldre's code? :/ If yes I really should have paid more attention during the review of its task ^^

I really don't like that `startValue` stuff I see in _renderCustomXML, seems really to be the point of the `_computeWidgetState` method. What if I have another option/widget which tweaks this feature (I don't know, something like a button which "clears all visibility dependencies of this snippet") -> I feel that the m2m widget would still show records even though that button was clicked since _computeWidgetState returns nothing?

If it is something related to SAD's code, could you maybe investigate a bit to see if this can be changed?

File: addons/website_forum/models/forum.py [hunk 36]
```diff
@@ -61,7 +61,9 @@ class Forum(models.Model):
                                 </div>
                             </div>
                         </div>
-                    </section>""")
+                    </section>""",
+        sanitize_attributes=False,
+        sanitize_form=False)
```
Comment (qsm-odoo) [inline addons/website_forum/models/forum.py:66]: You could do that in a separate commit of the same PR ;)

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Some new comments, especially interested in the m2m _computeWidgetState one :wink:

File: addons/website/static/src/js/editor/snippets.options.js (hunk 30 above)
Comment (detrouxdev) [reply addons/website/static/src/js/editor/snippets.options.js:2855]: The issue is that at some point in the [cleanForSave](https://github.com/odoo/odoo/blob/af9454361f88e024af85d43204964ae396b7de9b/addons/web_editor/static/src/js/editor/snippets.editor.js#L290) of the snippets.edditor, if an element has an o_snippet_invisible it will set dataset.invisible and do a bunch of operations if it detects that o_snippet_invisible is on the target. That's why I remove it on the cleanForSave of the option as I think those operations are meant for popups and can lead to a block staying permanently hidden.

Comment (detrouxdev) [reply addons/website/static/src/js/editor/snippets.options.js:2855]: After a better look into it, it doesn't seem like removing o_snippet_inivisible in cleanForSave prevents this from happening anyway. It doesn't seem like data-invisible impact the visibility.

File: addons/website/static/src/js/editor/snippets.options.js (hunk 35 above)
Comment (detrouxdev) [reply addons/website/static/src/js/editor/snippets.options.js]: It seems like I had badly interpreted a comment which said that the initial value does not come from the _computeWidgetState (because the widget will fetch records from the DB). If we return early however I believe the value will still come from _computeWidgetState. I have changed my code to use the _computeWidgetState.

**Review by Guillaume-gdi — COMMENTED** [review]
Here are some small comments. I did not go into depth but rather looked at the form.
I tried to apply the comments I had on my PR to yours. So maybe some of my comments are not right. If so, I'm interested to know why.

File: addons/website/static/src/js/editor/snippets.options.js [hunk 37]
```diff
@@ -2824,6 +2824,191 @@ options.registry.ReplaceImage = options.Class.extend({
...
+     *
+     * @override
+     */
+    selectDataAttribute(previewMode, widgetValue, params) {
+        this._super(...arguments);
+        if (params.attributeName === 'visibility') {
+            if (widgetValue === 'conditional') {
+                const collapseEl = this.$el.children('we-collapse')[0];
+                this._toggleCollapseEl(collapseEl);
+                this.trigger_up('snippet_option_visibility_update', { show: true });
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2983]: I am missing something in this code... is not this event supposed to work only for `data-invisible="1"` fields? I can't remember. But this may have something to do with your need to remove the d-none class in `cleanForSave` as I explained in my other comment: I have the feeling your snippet is always `data-invisible="1"` while it should only be when conditional visibility.

File: addons/website/static/src/js/editor/snippets.options.js [hunk 38]
```diff
@@ -2824,6 +2824,191 @@ options.registry.ReplaceImage = options.Class.extend({
...
+    // Private
+    //--------------------------------------------------------------------------
+
+    /**
+     * @override
+     */
+    async _computeWidgetState(methodName, params) {
+        if (methodName === 'selectRecord') {
+            // m2m value widget does not use computeWidgetState
+            return this.$target[0].dataset[params.attributeName] || '[]';
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:3006]: Hum, did not you forget to remove the comment?

File: addons/website/views/snippets/snippets.xml [hunk 39]
```diff
@@ -1060,6 +1060,80 @@
...
+                <we-button data-select-data-attribute="">Visible for</we-button>
+                <we-button data-select-data-attribute="hide">Hidden for</we-button>
+            </we-select>
+            <we-many2many t-if="request.session.geoip.get('country_code')"
+                            string=" " 
+                            data-dependencies="visibility_conditional" 
+                            data-attribute-name="visibilityValueCountry" 
+                            data-no-preview="true" 
+                            data-model="res.country" data-call-with="code" data-select-record="" data-fields="[&quot;code&quot;]" 
+                            data-allow-delete="true" data-fakem2m="true"/>
```
Comment (qsm-odoo) [inline addons/website/views/snippets/snippets.xml:1085]: This part of the UI is kinda repeated 4x times with a few different attributes -> make it a parametrized template that you can t-call 4x here, for each visibility condition part. For the t-if, you should not duplicate it but wrap the two widgets in a 
`<t t-if="..."></t>` (well, wrapping the t-call after the change I am asking here :wink:)

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
A few small changes required (I guess you could be done with them today) then we'll ask India for an urgent testing tomorrow so you can have the time to potentially fix the bugs they would have found before you are gone on Friday :+1:

**Comment by qsm-odoo** [comment]
@robodoo rebase-ff r+

**Comment by kig-odoo** [comment]
Hello @qsm-odoo @detrouxdev @JKE-be ,

Hope you're doing well.

This is really cool feature.
But I've noticed a bug or maybe you can consider it as an improvement.

I've noticed that the many2many widget for languages is fetching all active languages from the backend.
see:
image


Maybe instead of it, many2many widget should only fetch languages that are available for the website.
image

**Comment by qsm-odoo** [comment]
Hello @kig-odoo 

I am fine, hope you are doing well too :) Indeed, I'll create a task for it, thanks :+1:

**Comment by kig-odoo** [comment]
Hello @qsm-odoo

Thank you for your response.

Today I was exploring the code and I've found a few more bugs related to this feature.

**1.** [Here](https://github.com/odoo/odoo/blob/master/addons/website/static/src/js/editor/snippets.options.js#L3072) We are generating visibilityID by combining the records ids.

For example, my website has 3 languages. A(ID:1), B(ID:2) and C(ID:12)

I've 2 sections on my website page SECTION - 1, SECTION - 2

For SECTION - 1, I've configured visibility like "Hide for", Language (A AND B)
So the computed visibilityID will be `lang12h` 
And the selector will be like: `html[lang= .... [data-visibility-id="lang12h"],html[lang= .... [data-visibility-id="lang12h"]`

For SECTION - 2, I've configured visibility like "Hide for", Language (C)
So the computed visibilityID will be **same as SECTION - 1** `lang12h` 
And selector will be like: `html[lang="zh-TW"] body:not(.editor_enable) [data-visibility-id="lang12h"]`

**Result: Both SECTION - 1 and SECTION - 2 will be hidden for all three languages.**

I don't fully analyze the feature but maybe.
You should replace [Line](https://github.com/odoo/odoo/blob/1c442782f887a8c16bae05a43fae13a310ac05df/addons/website/static/src/js/editor/snippets.options.js#L3001)

Replace:
`+ records.reduce((acc, record) => acc += record.id, '')`
By:
`+ records.reduce((acc, record) => acc += '_' + record.id, '')`

**2.**

Regarding UTM based options on the snippet,selector will be based on field Campaigns name 
So the selector will be like: 
`html[data-utm-campaign="Sale"] body:not(.editor_enable) [data-visibility-id="utm-campaign_1h"]`

But note that  Campaigns name is the translatable field:
image

So in case, someone has generated a translatable term for the Campaigns.
Block will be hidden in one language and displayed in the second language

**Comment by JKE-be** [comment]
For .2

UTM should never have become translatable... But because they do it, now there are created a new way of 'identifier' :thinking: 
So wait and see. But btw the goal of this one is more to display dynamic content if you come from a facebook link e.g.. So you provide yourself the url and so the uniq um_campaign (untranslated) in this case.

**Comment by kig-odoo** [comment]
Right now, Sources and Campaigns names both are translated fields. 
Somehow Mediums is not translatable :smile: 

I think UTMs are just marketing and traceability. And mostly it will be in URLs or in cookies. And not visible to the enduser directly.
So yes, you're right, no need to make those fields to be translatable.

**Comment by qsm-odoo** [comment]
Thanks @kig-odoo, could have been a while to see the effect of this bug, best to have caught it before we freeze a version :+1: 

(1) being fixed at https://github.com/odoo/odoo/pull/75386

**Comment by kig-odoo** [comment]
Thanks Quentin for quick fix :)
